From 9b262d3dffb0cb6615f9e6804db8f85f58d9d02e Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Mon, 2 May 2022 14:23:58 +0200
Subject: [PATCH 37/39] i2c: stm32: add support for STM32MP25 soc

The stm32mp25 embeds a new FMP bit in the CR1 register,
necessary when running at Fast Mode Plus frequency.
This bit should be used instead of the SYSCFG bit used on other platforms.

Add a new compatible "st,stm32mp25-i2c" to distinguish between the SoCs
and a boolean "fmp_cr1_bit" within the setup structure in order to know
if the platform has the FMP bit within CR1 is available or not.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git 9abbd4a9743b82099251ddaeda99bbf4f9f5bc19 i2c: stm32: add support for STM32MP25 soc]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I905b6a8eede51d56f7dee17a59b2efdee2c1dc7d
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/250312
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
Tested-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
---
 drivers/i2c/stm32f7_i2c.c | 33 ++++++++++++++++++++++-----------
 1 file changed, 22 insertions(+), 11 deletions(-)

diff --git a/drivers/i2c/stm32f7_i2c.c b/drivers/i2c/stm32f7_i2c.c
index 3f51b1dd1db..b78405f322d 100644
--- a/drivers/i2c/stm32f7_i2c.c
+++ b/drivers/i2c/stm32f7_i2c.c
@@ -169,9 +169,11 @@ struct stm32_i2c_setup {
 /**
  * struct stm32_i2c_data - driver data for I2C configuration by compatible
  * @fmp_clr_offset: Fast Mode Plus clear register offset from set register
+ * @fmp_cr1_bit: Fast Mode Plus control is done via a bit in CR1
  */
 struct stm32_i2c_data {
 	u32 fmp_clr_offset;
+	bool fmp_cr1_bit;
 };
 
 /**
@@ -269,6 +271,10 @@ static const struct stm32_i2c_data stm32mp13_data = {
 	.fmp_clr_offset = 0x4,
 };
 
+static const struct stm32_i2c_data stm32mp25_data = {
+	.fmp_cr1_bit = true,
+};
+
 static int stm32_i2c_check_device_busy(struct stm32_i2c_priv *i2c_priv)
 {
 	struct stm32_i2c_regs *regs = i2c_priv->regs;
@@ -796,6 +802,8 @@ static int stm32_i2c_write_fm_plus_bits(struct stm32_i2c_priv *i2c_priv)
 	int ret;
 	bool enable = i2c_priv->speed > I2C_SPEED_FAST_RATE;
 
+	/* TODO STM32MP25: handle FMP bit in CR1  register (fmp_cr1_bit = true) */
+
 	/* Optional */
 	if (IS_ERR_OR_NULL(i2c_priv->regmap))
 		return 0;
@@ -931,19 +939,21 @@ static int stm32_of_to_plat(struct udevice *dev)
 
 	i2c_priv->setup.analog_filter = dev_read_bool(dev, "i2c-analog-filter");
 
-	/* Optional */
-	i2c_priv->regmap = syscon_regmap_lookup_by_phandle(dev,
-							   "st,syscfg-fmp");
-	if (!IS_ERR(i2c_priv->regmap)) {
-		u32 fmp[3];
+	if (!data->fmp_cr1_bit) {
+		/* Optional */
+		i2c_priv->regmap = syscon_regmap_lookup_by_phandle(dev,
+								   "st,syscfg-fmp");
+		if (!IS_ERR(i2c_priv->regmap)) {
+			u32 fmp[3];
 
-		ret = dev_read_u32_array(dev, "st,syscfg-fmp", fmp, 3);
-		if (ret)
-			return ret;
+			ret = dev_read_u32_array(dev, "st,syscfg-fmp", fmp, 3);
+			if (ret)
+				return ret;
 
-		i2c_priv->regmap_sreg = fmp[1];
-		i2c_priv->regmap_creg = fmp[1] + data->fmp_clr_offset;
-		i2c_priv->regmap_mask = fmp[2];
+			i2c_priv->regmap_sreg = fmp[1];
+			i2c_priv->regmap_creg = fmp[1] + data->fmp_clr_offset;
+			i2c_priv->regmap_mask = fmp[2];
+		}
 	}
 
 	return 0;
@@ -958,6 +968,7 @@ static const struct udevice_id stm32_i2c_of_match[] = {
 	{ .compatible = "st,stm32f7-i2c", .data = (ulong)&stm32f7_data },
 	{ .compatible = "st,stm32mp15-i2c", .data = (ulong)&stm32mp15_data },
 	{ .compatible = "st,stm32mp13-i2c", .data = (ulong)&stm32mp13_data },
+	{ .compatible = "st,stm32mp25-i2c", .data = (ulong)&stm32mp25_data },
 	{}
 };
 
-- 
2.53.0

